### PR TITLE
POC: asynchronous federated training on MNIST using websockets and jit

### PR DESCRIPTION
**Pull request by midokura-silvia**
This pull request contains code to train asynchronously on different workers. 
A concrete example is given with MNIST and the data being separated by digits on three different workers. 
The distribution of the model uses jit.trace() and the communication with the workers is a websocket connection. 
The example is located in the folder: examples/tutorials/advanced/websockets-example-MNIST-parallel

### CODE REVIEW

File: syft/federated/federated_client.py [hunk 1]
```diff
@@ -54,6 +57,17 @@ def _build_optimizer(self, optimizer_name: str, model, lr: float) -> th.optim.Op
         return self.optimizer
 
     def fit(self, dataset_key, **kwargs):
+        """Fit a model on the local dataset as specified in the local TrainConfig object
```
Comment (mari-linhares) [inline syft/federated/federated_client.py]: Missing period at the end of all sentences

File: syft/federated/federated_client.py [hunk 2]
```diff
@@ -64,28 +78,50 @@ def fit(self, dataset_key, **kwargs):
...
+                self.train_config.max_nr_batches
+                and iteration_count >= self.train_config.max_nr_batches
+            ):
+                logger.debug("Accuracy: %s", accuracy(output, target))
+                break
 
         return loss
+
+
+def accuracy(pred_softmax, target):
```
Comment (mari-linhares) [inline syft/federated/federated_client.py]: I don't think this should live here.

File: syft/federated/federated_client.py [hunk 3]
```diff
@@ -64,28 +78,50 @@ def fit(self, dataset_key, **kwargs):
...
-                data, target = self.datasets[dataset_key][data_indices]
-                self.optimizer.zero_grad()
-                output = model.forward(data)
-                loss = loss_fn(output, target)
-                loss.backward()
-                self.optimizer.step()
+        iteration_count = 0
+        for (data, target) in data_loader:
+
+            if iteration_count % 25 == 0:
```
Comment (mari-linhares) [inline syft/federated/federated_client.py]: Remove this logs or create a parameter called `iteration_count`.

File: syft/workers/websocket_client.py [hunk 4]
```diff
@@ -113,12 +120,41 @@ def list_objects_remote(self):
     def objects_count_remote(self):
         return self._send_msg_and_deserialize("objects_count")
 
-    def fit(self, dataset_key, **kwargs):
+    async def fit(self, dataset_key, **kwargs):
```
Comment (mari-linhares) [inline syft/workers/websocket_client.py]: this function is not easy to follow, maybe break this into multiple functions. Also, I think `synchronous_fit` should be called `fit` and this one called `async_fit`. Or even create a wrapper function `fit` with a parameter `async` with default value = False.

Comment (midokura-silvia) [reply syft/workers/websocket_client.py]: A wrapper function is not possible, as the whole chain of async functions needs the async keyword. 
But I can do the renaming.

File: examples/tutorials/advanced/websockets-example-MNIST-parallel/Asynchronous-federated-learning-on-MNIST.ipynb [hunk 5]
```diff
@@ -0,0 +1,1316 @@
+{
```
Comment (robert-wagner) [inline examples/tutorials/advanced/websockets-example-MNIST-parallel/Asynchronous-federated-learning-on-MNIST.ipynb]: <p>Should train be true for loading this data given it is for testing?</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2217/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

**Comment by mari-linhares** [comment]
Closing this in favor of #2289

**Pull request closed without merging**